targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_K66F/serial_api.c

```diff
@@ -294,49 +332,429 @@ void serial_set_flow_control(serial_t *obj, FlowControl type, PinName rxflow, Pi
+    serial_receive_asynch(obj);
+
+    /* Can't enter deep sleep as long as UART transfer is active */
+    sleep_manager_lock_deep_sleep();
+}
+
+uint8_t serial_tx_active(serial_t *obj)
+{
+    if (obj->serial.txstate == kUART_TxIdle) {
+        return 0;
+    }
+
+    return 1;
+}
+
+uint8_t serial_rx_active(serial_t *obj)
+{
+    if (obj->serial.rxstate == kUART_RxIdle) {
+        return 0;
+    }
+
+    return 1;
+}
+
+int serial_irq_handler_asynch(serial_t *obj)
+{
+    int status = 0;
+    //uint8_t *buf = (uint8_t*)obj->rx_buff.buffer;
+    uint32_t status_flags = UART_GetStatusFlags(uart_addrs[obj->serial.index]);
+
+    /* Determine whether the current scenario is DMA or IRQ, and act accordingly */
+    if (obj->serial.uartDmaRx.dmaUsageState == DMA_USAGE_ALLOCATED || obj->serial.uartDmaRx.dmaUsageState == DMA_USAGE_TEMPORARY_ALLOCATED) {
+        /* DMA implementation */
+        if ((obj->serial.txstate != kUART_TxIdle)  && (obj->serial.uart_dma_handle.txState == kUART_TxIdle)) {
+            obj->serial.txstate = kUART_TxIdle;
+            status |= SERIAL_EVENT_TX_COMPLETE;
+            /* Transmit is complete, re-enable entry to deep sleep mode */
+            sleep_manager_unlock_deep_sleep();
+        }
+
+        if ((obj->serial.rxstate != kUART_RxIdle) && (obj->serial.uart_dma_handle.rxState == kUART_RxIdle)) {
+            obj->serial.rxstate = kUART_RxIdle;
+            status |= SERIAL_EVENT_RX_COMPLETE;
+            /* Receive is complete, re-enable entry to deep sleep mode */
+            sleep_manager_unlock_deep_sleep();
+        }
+
+        /* Release the dma channels if they were opportunistically allocated */
+        if (obj->serial.uartDmaRx.dmaUsageState == DMA_USAGE_TEMPORARY_ALLOCATED) {
+            /* Ensure both TX and RX channels are idle before freeing them */
+            if ((obj->serial.uart_dma_handle.txState == kUART_TxIdle) && (obj->serial.uart_dma_handle.rxState == kUART_RxIdle)) {
+                dma_channel_free(obj->serial.uartDmaRx.dmaChannel);
+                dma_channel_free(obj->serial.uartDmaTx.dmaChannel);
+                obj->serial.uartDmaRx.dmaUsageState = DMA_USAGE_OPPORTUNISTIC;
+            }
+        }
+    } else {
+        /* Interrupt implementation */
+        if ((obj->serial.txstate != kUART_TxIdle) && (obj->serial.uart_transfer_handle.txState == kUART_TxIdle)) {
+            obj->serial.txstate = kUART_TxIdle;
+            status |= SERIAL_EVENT_TX_COMPLETE;
+            /* Transmit is complete, re-enable entry to deep sleep mode */
+            sleep_manager_unlock_deep_sleep();
+        }
+
+        if ((obj->serial.rxstate != kUART_RxIdle) && (obj->serial.uart_transfer_handle.rxState == kUART_RxIdle)) {
+            obj->serial.rxstate = kUART_RxIdle;
+            status |= SERIAL_EVENT_RX_COMPLETE;
+            /* Receive is complete, re-enable entry to deep sleep mode */
+            sleep_manager_unlock_deep_sleep();
+        }
+    }
+#if 0
```

Comment:
can this be removed?

Reply:
This feature is not complete and needs to be completed. We can leave it in for now and I will work to update this support to support CHARACTER_MATCH